Add Datadog exporter overall structure

mx-psi opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Add Datadog exporter overall structure following the [contributing guidelines](https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component).

This is a rework of #900: instead of using `statsd` it will report directly to Datadog's backend. It incorporates the feedback from that PR's reviews.

**Link to tracking Issue:** n/a

**Testing:** Added unit tests, tested full version (available on [our fork](https://github.com/DataDog/opentelemetry-collector-contrib)) on an end to end testing environment.

**Documentation:** This adds the README for the exporter with the intended instructions for the full metrics exporter.

**Note**: The commit numbering refers to the PRs on our fork but are shown incorrectly by Github.

mx-psi committed:
Add DataDog exporter back from old fork
commit 99129fb96e29e9c1a92da00b7e3f8efcae8a31e8
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Thu Sep 3 18:10:28 2020 +0200

    Handle namespace at initialization time

commit babca25927926a60c0c416294af3aadf784d41b9
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Thu Sep 3 17:23:53 2020 +0200

    Initialize on a separate function
    This way the variables can be checked without worrying about the env

commit 24d0cb4cc566fa5313a8650c904a27bea68bf555
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Thu Sep 3 14:30:35 2020 +0200

    Check environment variables for unified service tagging

commit 6695f8297ab8b1fcae71b05acb027c4a0992e3a0
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Wed Sep 2 14:57:37 2020 +0200

    Add support for sending metrics through the API
    - Use datadog.Metric type for simplicity
    - Get host if unset

commit c3666031b63710e61e761e7fdba8cbc65bcb78de
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Wed Sep 2 09:56:24 2020 +0200

    Disable Queue and Retry settings (#72)

    These are handled by the statsd package.
    OpenTelemetry docs are confusing and the default configuration (disabled)
    is different from the one returned by "GetDefault..." functions

commit a660b56ab3f6a52c7aa87f9d87eafa387af1c76c
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Tue Sep 1 15:26:14 2020 +0200

    Add support for summary and distribution metric types (#65)

    * Add support for summary metric type

    * Add support for distribution metrics

    * Refactor metrics construction
    - Drop name in Metrics (now they act as Metric values)
    - Refactor constructor so that errors happen at compile-time

    * Report Summary total sum and count values
    Snapshot values are not filled in by OpenTelemetry

    * Report p00 and p100 as `.min` and `.max`
    This is more similar to what we do for our own non-additive type

    * Keep hostname if it has not been overridden

commit c95adc46e4a2b60c205d0f17e70aa3d610d227c3
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Thu Aug 27 13:00:02 2020 +0200

    Update dependencies and `make gofmt`
    The collector was updated to 0.9.0 upstream

commit 20afb0e9e9f2afdd987740850f47dbe6471d69ef
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Wed Aug 26 18:25:49 2020 +0200

    Refactor configuration (#45)

    * Refactor configuration
    * Implement telemetry and tags configuration handling
    * Update example configuration and README file
    Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

commit fdc98b57a9e04c3f275087167af35d8c3d217184
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Fri Aug 21 11:09:08 2020 +0200

    Initial DogStatsD implementation (#15)

    Initial metrics exporter through DogStatsD with support for all metric types but summary and distribution

commit e848a6054bd21d45763e213c99d8c80197246cb8
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Fri Aug 21 10:42:45 2020 +0200

    Bump collector version

commit 58be9a47cfba27ca1fa199765b50dba7fea18445
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Thu Aug 6 10:04:32 2020 +0200

    Address linter

commit 695430c585bd7dd34b499192d5b5f9f2d79b4ae4
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Tue Aug 4 13:28:01 2020 +0200

    Fix field name error

    MetricsEndpoint was renamed to MetricsURL

commit 168b3194c1109dc195f92999384bf7569b629178
Author: Pablo Baeyens <pablo.baeyens@datadoghq.com>
Date:   Mon Aug 3 11:05:01 2020 +0200

    Create initial outline for Datadog exporter (#1)

    * Add support for basic configuration options
    * Documents configuration options

mx-psi committed:
go mod tidy

mx-psi committed:
Address feedback from upstream PR we did not merge (#1)

* Backport changes from upstream PR

Remove `err` from MapMetrics

* Remove usage of pdatautil

* Fix tests

* Use TCPAddr

* Review which functions should be private

mx-psi committed:
Remove DogStatsD mode (#2)

* Remove DogStatsD mode

* go mod tidy

* Remove mentions to DogStatSD

mx-psi committed:
Improve test coverage (#3)

* Improve test coverage

Added unit tests for
- API key censoring
- Hostname
- Metrics exporter

Renamed test and implementation files for consistency

* Add one additional test

mx-psi committed:
Remove client validation (#6)

The zorkian API does not validate the API key unless you also have
an application key, even though the endpoint works without it.

I am removing this validation until this gets fixed on the zorkian library

mx-psi committed:
Keep only configuration and factory methods

Following the contribution guidelines we need to make a first PR
with this

mx-psi committed:
Merge remote-tracking branch 'upstream/master' into exporter-config

mx-psi committed:
Use latest version of collector

mx-psi committed:
Remove `report_percentiles` option
It is not being used as of now until the OTLP metrics format
stabilizes and we have a Summary type metric again

mx-psi committed:
Correct configuration

The API key is now a required setting

mx-psi committed:
Remove test not relevant for this PR

mx-psi committed:
Remove unnecessary imports after removing test

bogdandrutu commented on exporter/datadogexporter/config.go:
```diff
@@ -0,0 +1,170 @@
+
+// UpdateWithEnv gets the unified service tagging information
+// from the environment variables.
+func (t *TagsConfig) UpdateWithEnv() {
```
We already support loading env variables, and we would expect you to use the same mechanism. You can simply make the default `Hostname = "${DD_HOST}"`.

So I would suggest this custom mechanism to be removed

mx-psi replied:
I changed it to use the existing mechanism, thanks for the comment!

mx-psi committed:
Address review comment

tigrannajaryan commented on exporter/datadogexporter/config.go:
```diff
@@ -0,0 +1,145 @@
+)
+
+var (
+	errUnsetAPIKey = errors.New("the Datadog API key is unset")
```
Nit: it may be more user friendly to mention the actual config setting that is missing:
```suggestion
	errUnsetAPIKey = errors.New("api.key is not set")
```

tigrannajaryan reviewed (approved):
LGTM. Can be merged if other approvers don't have comments.

tigrannajaryan commented:
Please resolve the conflict.

mx-psi committed:
Merge branch 'master' into exporter-config

mx-psi committed:
Apply suggestions from code review

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

mx-psi committed:
Separate documentation into two examples
One example with the minimal configuration, for sending to `datadoghq.com`
and a second one for sending to `datadoghq.eu`

mx-psi commented on exporter/datadogexporter/README.md:
```diff
@@ -0,0 +1,17 @@
+datadog:
+  api:
+    key: "<API key>"
+    # site: datadoghq.eu for sending data to the Datadog EU site
```
I reworded it into two examples so that this minimal example is nicer

mx-psi commented:
@tigrannajaryan I addressed the last comments and resolved the conflict.
